src/threshold-ts/staking/index.ts

```diff
@@ -249,21 +266,109 @@ export class Staking implements IStaking {
+    )
+      .map((event) => ({
+        stakingProvider: event.args?.stakingProvider,
+        stakeType: event.args?.stakeType,
+      }))
+      .reverse()
+
+    const { current, outdated } = await this.findRefreshedKeepStakes(owner)
+
+    const stakingProviders = stakes
+      .filter(({ stakingProvider }) => !outdated.includes(stakingProvider))
+      .concat(
+        current.map((stakingProvider) => ({
+          stakingProvider,
+          stakeType: StakeType.KEEP,
+        }))
+      )
 
     return await Promise.all(
-      stakedEvents.map((stakedEvent) =>
-        this.getStakeByStakingProvider(
-          stakedEvent.args?.stakingProvider,
-          stakedEvent.args?.stakeType
-        )
+      stakingProviders.map(({ stakingProvider, stakeType }) =>
+        this.getStakeByStakingProvider(stakingProvider, stakeType)
       )
     )
   }
+
+  findRefreshedKeepStakes = async (
+    owner: string
+  ): Promise<OwnerRefreshedResult> => {
+    // Find all events where the `owner` was set as a new owner or old owner of
+    // the stake.
+    const ownerRefreshedEventsFilteredByNewOwner = await getContractPastEvents(
+      this._staking,
+      {
+        eventName: "OwnerRefreshed",
+        filterParams: [null, null, owner],
+        fromBlock: this.STAKING_CONTRACT_DEPLOYMENT_BLOCK,
+      }
+    )
+
+    const ownerRefreshedEventsFilteredByOldOwner = await getContractPastEvents(
+      this._staking,
+      {
+        eventName: "OwnerRefreshed",
+        filterParams: [null, owner, null],
+        fromBlock: this.STAKING_CONTRACT_DEPLOYMENT_BLOCK,
+      }
+    )
+
+    const ownerRefreshedEvents = [
+      ...ownerRefreshedEventsFilteredByNewOwner,
+      ...ownerRefreshedEventsFilteredByOldOwner,
+    ].sort((a, b) => a.blockNumber - b.blockNumber)
+
+    // Convert to `Set` to remove duplicated staking provider addresses(the
+    // owner can be changed multiple times for the same staking provider) and
+    // then to an array so we can use `map`, `filter` and other array methods.
+    const possibleStakingProviders: string[] = Array.from(
+      new Set(ownerRefreshedEvents.map((event) => event?.args?.stakingProvider))
+    )
+
+    const multicalls: ContractCall[] = possibleStakingProviders.map(
+      (stakingProvider) => ({
+        address: this._staking.address,
+        interface: this._staking.interface,
+        method: "rolesOf",
+        args: [stakingProvider],
+      })
+    )
+
+    // For each staking provider, we need to verify that the current owner of
+    // the stake is indeed the one we are looking for (the `owner` argument).
+    // It's possible that the owner can be changed multiple times. Instead of
+    // iterating through the `OwnerRefreshed` events and comparing the
+    // `oldOwner` and `newOwner` params from that event, we can just check the
+    // current owner for a given staking provider by calling `rolesOf`.
+    const rolesOf: { owner: string }[] = await this._multicall.aggregate(
```

Comment:
I think we can use [Role Of interface](https://github.com/threshold-network/token-dashboard/blob/main/src/threshold-ts/staking/index.ts#L40) here.

```suggestion
    const rolesOf: RolesOf[] = await this._multicall.aggregate(
```

Reply:
2a2d714